llvm/include/llvm/IR/RepoDefinition.h

```diff
@@ -107,56 +103,116 @@ struct GODigestState {
   /// A reference to an object's dependencies. The GOInfo struct takes ownership
   /// of it.
   const GOVec &Dependencies;
-  /// A bool which is true if GO's hash value has been changed.
-  bool Changed;
   GODigestState() = delete;
-  GODigestState(const GOVec &Contributions, const GOVec &Dependencies,
-                bool Changed)
-      : Contributions(Contributions), Dependencies(Dependencies),
-        Changed(Changed) {}
+  GODigestState(const GOVec &Contributions, const GOVec &Dependencies)
+      : Contributions(Contributions), Dependencies(Dependencies) {}
 };
 
-/// Calculate the initial hash value, dependencies and contributions for the
-/// global object 'G'.
-/// \param G Calculated global object.
-/// \param GOI a DenseMap containing the global object information.
-///
-template <typename GlobalType>
-void calculateGOInfo(const GlobalType *G, GOInfoMap &GOI) {
-  GOInfo Result = calculateDigestAndDependenciesAndContributedToGVs(G);
-  // ContributedToGVs of an object is used to update other objects'
-  // contributions. For example, if the ContributedToGVs of function `foo`  are
-  // global variables `g` and `q`, the function `foo` is in the contributions of
-  // `g` and `q`.
-  // ContributedToGVs[`foo`] = [`g`, `q`]
-  // ====> Contributions[`g`] = [`foo`],
-  //       Contributions[`q`] = [`foo`]
-  for (auto &GO : Result.Contributions) {
-    assert(isa<GlobalVariable>(GO) &&
-           "Only global variables can have contributions!");
-    assert(isa<llvm::Function>(G) && "All contributions are functions!");
-    GOI[GO].Contributions.emplace_back(G);
+/// Helper Class for generating the hash for each global objects on the Module.
+/// Also caches hashes for previously generated global objects (GOs).
+class ModuleHashGenerator {
+
+  enum class Tags : char {
+    Backref = 'R',
+    End = 'E',
+    GO = 'T',
+  };
+
+  /// A map of GO to a unique number in the dependencies graph.
+  GOStateMap Visited;
+  /// Used to keep the GO information.
+  GOInfoMap GOIMap;
+  /// Memoize GO's hash if one of the following conditions is satisfied:
+  /// 1. GO is not in a cycle/loop.
+  /// 2. GO is in a self-loop but not if there is a loop involving another
+  /// object.
+  MemoizedHashes GOHashCache;
+  /// True if the module M is changed.
+  bool Changed = false;
+
+public:
+  ModuleHashGenerator() {}
+  /// calculate the digest for all global objects inside of the Module M.
+  void digestModule(Module &M);
+
+  /// Compute the GO's hash value using the digest of GO's dependencies if it is
+  /// stored in the RepoDefinition metadata.
+  /// \param GO The global object whose digest is to be computed.
+  /// \return The global object's digest value.
+  DigestType calculateDigestUseRepoDefinition(const GlobalObject *GO);
```

Comment:
Would digestObject (or digestObjectUseRepoDefinition) be a better name for this (to be consistent with digestModule())?

Reply:
Changed, thanks.